[DOC] Updates query syntax (#3291)
RediSearch/RediSearch: docs/docs/reference/Query_Syntax.md

linkTitle: "Query syntax"
weight: 1
description: >
    Learn how to use search query syntax
---

You can use simple syntax for complex queries using these rules:

* Multiword phrases are lists of tokens, for example, `foo bar baz`, and imply intersection (AND) of the terms.
* Exact phrases are wrapped in quotes, for example, `"hello world"`.

> * `OR` unions are expressed with a pipe (`|`), for example, `hello|hallo|shalom|hola`.

Reviewer: There are related fixes in DIALECT 2
See "Examples of queries impacted" in [v2.4.3](https://github.com/RediSearch/RediSearch/releases/tag/v2.4.3/) Release Notes

> Consider a simple query with negation -hello world.
> ...

> Another example illustrating the differences in parser behavior
> hello world | "goodbye" moon.
> 
> In DIALECT 1, this query is interpreted as searching for (hello world | "goodbye") moon.
> In DIALECT 2, this query is interpreted as searching for either hello world OR "goodbye"  moon
> 
>

* Exact phrases are wrapped in quotes, for example, `"hello world"`.
* `OR` unions are expressed with a pipe (`|`), for example, `hello|hallo|shalom|hola`. 

  {{% alert title="Note" color="warning" %}}
  Consider the differences in parser behavior in example `hello world | "goodbye" moon`:
  * In DIALECT 1, this query is interpreted as searching for `(hello world | "goodbye") moon`.
  * In DIALECT 2 or greater, this query is interpreted as searching for either `hello world` **OR** `"goodbye" moon`.
  {{% /alert %}}
  
* `NOT` negation of expressions or subqueries is expressed with a subtraction symbol (`-`), for example, `hello -world`. Purely negative queries (for example, `-foo` or `-@title:(foo|bar)`) are also supported.

  {{% alert title="Note" color="warning" %}}
  Consider a simple query with negation `-hello world`:
  * In DIALECT 1, this query is interpreted as "find values in any field that does not contain `hello` **AND** does not contain `world`". The equivalent is `-(hello world)` or `-hello -world`.
  * In DIALECT 2, this query is interpreted `as -hello` **AND** `world` (only `hello` is negated).

>   * In DIALECT 2, You could achieve the default behavior as in dialect 1 by updating your query to `-(hello world)`.

Author: ```suggestion
  * In DIALECT 2, to achieve the default behavior of dialect 1, update your query to `-(hello world)`.
```

* Multiword phrases are lists of tokens, for example, `foo bar baz`, and imply intersection (AND) of the terms.
* Exact phrases are wrapped in quotes, for example, `"hello world"`.
* `OR` unions are expressed with a pipe (`|`), for example, `hello|hallo|shalom|hola`. 

  {{% alert title="Note" color="warning" %}}
  Consider the differences in parser behavior in example `hello world | "goodbye" moon`:
  * In DIALECT 1, this query is interpreted as searching for `(hello world | "goodbye") moon`.

>   * In DIALECT 2, this query is interpreted as searching for either `hello world` **OR** `"goodbye" moon`.

Reviewer: Also in other occurrences
```suggestion
  * In DIALECT 2 or greater, this query is interpreted as searching for either `hello world` **OR** `"goodbye" moon`.
```

  {{% alert title="Note" color="warning" %}}
  Consider a simple query with negation `-hello world`:
  * In DIALECT 1, this query is interpreted as "find values in any field that does not contain `hello` **AND** does not contain `world`". The equivalent is `-(hello world)` or `-hello -world`.
  * In DIALECT 2 or greater, this query is interpreted `as -hello` **AND** `world` (only `hello` is negated).

>   * In DIALECT 2 or greater, to achieve the default behavior of dialect 1, update your query to `-(hello world)`.

Author: ```suggestion
  * In DIALECT 2 or greater, to achieve the default behavior of DIALECT 1, update your query to `-(hello world)`.
```

To specify which fields the query matches, prepend the expression with the `@` symbol, the field name, and a `:` (colon) symbol, per query expression or subexpression.

If a field modifier precedes multiple words or expressions, it applies only to the adjacent expression with DIALECT 1. With DIALECT 2 or greater, you extend the query to other fields.

Consider this simple query: `@name:James Brown`. Here, the field modifier `@name` is followed by two words: `James` and `Brown`.

* In DIALECT 1, this query would be interpreted as "find `James Brown` in the `@name` field".
* In DIALECT 2 or greater, this query will be interpreted as "find `James` in the `@name` field **AND** `Brown` in **ANY** text field. In other words, it would be interpreted as `(@name:James) Brown`.

> * In DIALECT 2 or greater, you can achieve the default behavior as in DIALECT 1 by updating your query to `@name:(James Brown)`.

Author: ```suggestion
* In DIALECT 2 or greater, to achieve the default behavior of DIALECT 1, update your query to `@name:(James Brown)`.
```
